events/reporter.go

```diff
@@ -0,0 +1,430 @@
+package events
+
+import (
+	"errors"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/timesync"
+)
+
+// reporter is the event reporter singleton.
+var reporter *EventReporter
+
+// ReportNewTx dispatches incoming events to the reporter singleton
+func ReportNewTx(tx *types.Transaction) {
+	Publish(NewTx{
+		ID:          tx.ID().String(),
+		Origin:      tx.Origin().String(),
+		Destination: tx.Recipient.String(),
+		Amount:      tx.Amount,
+		Fee:         tx.Fee,
+	})
+	ReportTxWithValidity(tx, true)
+}
+
+// ReportTxWithValidity reports a tx along with whether it was just invalidated
+func ReportTxWithValidity(tx *types.Transaction, valid bool) {
+	txWithValidity := TransactionWithValidity{
+		Transaction: tx,
+		Valid:       valid,
+	}
+	if reporter != nil {
+		if reporter.blocking {
```

Comment:
should this be thread safe? accessing reporter properties

Reply:
The properties are read only so I think we're good here, but you're right re: thread safety and setting the singleton reporter in the first place. In production code there should only ever be a single instance and it should only ever be initialized/torn down once, but it can't hurt to make it thread safe just in case -- will fix this.

Reply:
how is this? https://github.com/spacemeshos/go-spacemesh/pull/2071/commits/8a94e2dd8214359d182a9f2ceb94b87dd8725917